torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -25,722 +26,833 @@ struct propagation_error : std::exception {};
+      size_t idx1,
+      size_t idx2) {
+    auto expected_size =
+        at::infer_size(types[idx1]->sizes(), types[idx2]->sizes());
+    auto broadcast = [&](size_t input_idx) {
+      CompleteTensorTypePtr input_type = types.at(input_idx);
+      if (input_type->sizes() == expected_size)
+        return;
+      auto graph = node->owningGraph();
+      WithInsertPoint point_guard{node};
+      Node* expand = graph
+                         ->create(
+                             aten::expand,
+                             {node->inputs().at(input_idx),
+                              graph->insertConstant(expected_size),
+                              graph->insertConstant(false)})
+                         ->insertBefore(node);
+      PropagateShapeOnNode(expand);
+      node->replaceInput(input_idx, expand->output());
+    };
+    broadcast(idx1);
+    broadcast(idx2);
+    types[0] = node->inputs().at(idx1)->type()->expect<CompleteTensorType>();
+    types[1] = node->inputs().at(idx2)->type()->expect<CompleteTensorType>();
+  }
 
-void broadcastBinary(Node *node, std::vector<CompleteTensorTypePtr>& types, size_t idx1, size_t idx2) {
-  auto expected_size = at::infer_size(types[idx1]->sizes(), types[idx2]->sizes());
-  auto broadcast = [&](size_t input_idx) {
-    CompleteTensorTypePtr input_type = types.at(input_idx);
-    if (input_type->sizes() == expected_size)
-      return;
-    auto graph = node->owningGraph();
-    WithInsertPoint point_guard { node };
-    Node *expand = graph->create(aten::expand,
-                                 {node->inputs().at(input_idx),
-                                  graph->insertConstant(expected_size),
-                                  graph->insertConstant(false)})
-                        ->insertBefore(node);
-    PropagateShapeOnNode(expand);
-    node->replaceInput(input_idx, expand->output());
+  OperatorSet cannot_propagate_shape_by_running_it = {
+      "aten::gesv(Tensor self, Tensor A) -> (Tensor, Tensor)",
+      "aten::inverse(Tensor self) -> Tensor",
   };
-  broadcast(idx1);
-  broadcast(idx2);
-  types[0] = node->inputs().at(idx1)->type()->expect<CompleteTensorType>();
-  types[1] = node->inputs().at(idx2)->type()->expect<CompleteTensorType>();
-}
 
-bool isValidArgumentForRunning(Value* v) {
-  // allow constants
-  if(toIValue(v))
-    return true;
-  if(CompleteTensorTypePtr tt = v->type()->cast<CompleteTensorType>()) {
-    return !at::isIntegralType(tt->scalarType());
-  }
-  return v->type()->isSubtypeOf(FloatType::get());
-}
-bool isValidReturnForRunning(Value* v) {
-  return v->type()->isSubtypeOf(DynamicType::get()) ||
-      v->type()->isSubtypeOf(NumberType::get());
-}
 
-OperatorSet cannot_propagate_shape_by_running_it = {
-  "aten::gesv(Tensor self, Tensor A) -> (Tensor, Tensor)",
-  "aten::inverse(Tensor self) -> Tensor",
-};
+  // Check if this node relies on a value that has been
+  std::unordered_map<Node*, bool> dependsOnMutationMemo_;
```

Comment:
This is the main change: we disallow running a node to do shape propagation if it's value was clobbered by a mutation previously. Since we don't know whether that mutation was actually executed, it's not safe to run this node.